Implement welcome guide modal

noisysocks opened the pull request:
Closes #16315.

Implements the welcome guide modal described in https://github.com/WordPress/gutenberg/issues/16315#issuecomment-523174068.

![guide](https://user-images.githubusercontent.com/612155/69555384-6366f880-0fa3-11ea-9f58-a65f5817e936.gif)

## What's changed 

- There's a new `Guide` component in `@wordpress/nux` which implements the guide modal in a reusable way. This is marked experimental so that we can potentially move this to `@wordpress/components` in a follow-up PR.
- `WelcomeGuide` in `@wordpress/edit-post` will use `Guide` to display a welcome guide for the block editor if the user has not previously disabled tips. This replaces all of our instances of the `DotTip` component. We can potentially move away from using `hasDisabledTips` in a follow-up PR.
- There's a new _Welcome Guide_ button in the More menu which allows the user to open the guide again. This replaces the _Enable Tips_ setting in the Options modal.

## How to test

Unit and E2E tests are included. To manually test the upgrade process, follow these steps:

1. On `master`, open the block editor and ensure that you have dismissed tips.
1. Check out this branch and open the block editor. The welcome guide should not appear.
1. [Rest your browser's `localStorage`](https://developers.google.com/web/tools/chrome-devtools/storage/localstorage) to simulate being a new user.
1. Refresh the page. The guide should appear.
1. Dismiss or complete the guide.
1. Refresh the page. The guide should no longer appear.

noisysocks committed:
Implement welcome guide modal

noisysocks commented:
There's still lots to do here, but it would be good to get some early design feedback 🙂

noisysocks commented on packages/edit-post/src/components/welcome-guide-modal/index.js:
```diff
@@ -0,0 +1,65 @@
+					<BlockLibraryImage />
+					<div className="edit-post-welcome-guide-modal__text">
+						<h1>{ __( 'Get to know the block library' ) }</h1>
+						<p dangerouslySetInnerHTML={ {
```
https://github.com/WordPress/gutenberg/pull/17376 will let us get rid of this hack.

enriquesanchez commented:
Hey @noisysocks! 👋

Happy to help with design feedback. How can I get this to show up on my local install?

noisysocks commented:
Thanks @enriquesanchez! You can check out the branch and then open the Welcome Guide from the More menu:

<img width="277" alt="Screen Shot 2019-11-14 at 15 16 20" src="https://user-images.githubusercontent.com/612155/68826382-c8903380-06f1-11ea-95a9-cda130191c81.png">

noisysocks commented on packages/edit-post/src/components/welcome-guide-modal/index.js:
```diff
@@ -0,0 +1,70 @@
+	return (
+		<Modal
+			className="edit-post-welcome-guide-modal"
+			shouldCloseOnClickOutside={ false }
```
I've disabled this to work around an issue where the modal is immediately dismissed on page load because the post's title field receives focus. Let me know if you can think of a better alternative!

noisysocks commented on packages/edit-post/src/components/welcome-guide-modal/index.js:
```diff
@@ -0,0 +1,70 @@
+					</h1>
+					<BlockLibraryImage className="edit-post-welcome-guide-modal__image" />
+					<p className="edit-post-welcome-guide-modal__text">
+						{ __experimentalCreateInterpolateElement(
```
Works great, @nerrad! 😍

noisysocks commented on packages/edit-post/src/components/welcome-guide-modal/index.js:
```diff
@@ -0,0 +1,70 @@
+import { CanvasImage, EditorImage, BlockLibraryImage, InserterIconImage } from './vectors';
+
+export default function WelcomeGuideModal() {
+	const areTipsEnabled = useSelect( ( select ) => select( 'core/nux' ).areTipsEnabled() );
```
Using this selector from `core/nux` ensures that users who disabled tips will not get the intro modal when updating to the version of Gutenberg that includes this PR. 

Ideally we'd migrate this setting to `core/edit-post`, but I think it's best to do that if/when we remove the `core/nux` package.

noisysocks replied:
Let's do this in a follow-up PR.

aduth commented on packages/edit-post/src/components/welcome-guide-modal/index.js:
```diff
@@ -0,0 +1,70 @@
+						{ __experimentalCreateInterpolateElement(
+							__( 'All of the blocks available to you live in the Block Library. You’ll find it wherever you see the <InserterIconImage /> icon.' ),
+							{
+								InserterIconImage: (
```
I like how you chose to name the property to align with the component name, but in-so-doing it occurs to me whether we ought to have any coding guidelines around the names we choose for these interpolated keys? cc @nerrad

nerrad replied:
I think this is a question best answered by those working more closely with translators.  

One of the intents behind the interface is to provide additional guidance to translators on what a tag represents and signal that it shouldn't be translated.  With WordPress core in particular, I think generally translators know to leave tags alone (although that has been a smaller subset of tags I think).   

I think others might be better equipped to detail what standards we should have.  cc @ocean90 , @swissspidy  or @akirk ,  do you have any guidance here?

akirk replied:
I am huge proponent of "speaking" component placeholders. It's a very easy way to convey to translators what is being substituted here. `InserterIconImage` sounds good to me.

How and where would be put coding guidelines for this?

gziolo replied:
I like this proposal. Assuming that it's copied to all translations it would work great 👍 

For the full picture, there are some drawbacks discussed in the PR which test this new API. The discussion triggered by @johnbillion starts with this comment: https://github.com/WordPress/gutenberg/pull/18623#issuecomment-555995121.

gziolo replied:
@akirk do you suggest that the following:

```js
{ __experimentalCreateInterpolateElement(
	__( 'While writing, you can press <kbd>/</kbd> to quickly insert new blocks.' ),
	{ kbd: <kbd /> }
) }
```

could be easier refactored to:
```js
```js
{ __experimentalCreateInterpolateElement(
	__( 'While writing, you can press <Shortcut> to quickly insert new blocks.' ),
	{ Shortcut: <kbd>/</kbd> }
) }
```

TimothyBJacobs commented:
Would it be possible for the `<Guide />` component to be a part of NUX? This looks like something that'd be awesome to use on plugin admin screens as well. But since `<Guide />` is in `editPost` it'd be difficult to use.

----

Separately, it looks like #12632 would still be effected by this tips redesign. AFAICT, the close button here is dismissing _all_ guides instead of just the current "guide".

noisysocks commented:
> Would it be possible for the `<Guide />` component to be a part of NUX? This looks like something that'd be awesome to use on plugin admin screens as well. But since `<Guide />` is in `editPost` it'd be difficult to use.

Good suggestion! I'll look into this.

@mapk: Do you think a guide component would be useful outside of WP Admin? If so, then I think we should consider putting `Guide` and/or `GuideModal` into `@worpdress/components` so that we get to use the existing infrastructure there e.g. storybooks.

> Separately, it looks like #12632 would still be effected by this tips redesign. AFAICT, the close button here is dismissing _all_ guides instead of just the current "guide".

Thanks for pointing me to #12632. I'll look into moving away from `areTipsEnabled()`. My preference is to make `Guide` a dumb component and leave it up to the plugin developer to decide when to show it.

mapk commented:
> @mapk: Do you think a guide component would be useful outside of WP Admin? If so, then I think we should consider putting Guide and/or GuideModal into @worpdress/components so that we get to use the existing infrastructure there e.g. storybooks.

YES! Adding this to `@wordpress/components` sounds reasonable to me. Allowing other developers to use it could be amazing.

mapk commented:
I really like how this is shaping up! Really great work. Here's some design feedback:

### 1. First image needs to be centered better.
The webpage in the graphic feels like it's not quite centered. Even the blocks in the webpage don't look centered.

![Screen Shot 2019-11-20 at 8 34 45 AM](https://user-images.githubusercontent.com/617986/69261841-2cde4780-0b77-11ea-9257-0d681c4978f9.png)

### 2. First two graphics should be equal width.
I think this may help visually if both graphics in the first two images are equal width within the blue container.

![1](https://user-images.githubusercontent.com/617986/69261854-34055580-0b77-11ea-9a47-b6e1a3019df6.png)


### 3. Should we capitalize certain things?
"Block Editor" or "Block Library" seem like proper nouns. For example, we capitalize Block Library in the sentence, but not in the heading of the third screen.

### 4. Second screen text addition
Should we mention that you can move the block toolbar to be fixed at the top of the screen for less visual distractions?

### 5. This icon can probably use some visual cleanup.
Also lets make the triangle white to match the popover graphic too. Right now it's grey.

![Screen Shot 2019-11-20 at 9 29 02 AM](https://user-images.githubusercontent.com/617986/69262377-403de280-0b78-11ea-9922-0f7536c67e44.png)

I think these last few changes will dial it in beautifully.

enriquesanchez commented:
Hey @noisysocks! This is looking great! 🎉

Visuals
---

Would it be possible to align the close button so it has the same space around like the previous and next/get-started buttons?

<img width="841" alt="Screen Shot 2019-11-20 at 16 06 24" src="https://user-images.githubusercontent.com/3276087/69283207-9f6c1900-0bb1-11ea-9811-99d79b0d76e0.png">

I also noticed that the hover state of the page control buttons looks a bit off. It looks like the right border is missing half a pixel (I'm on a retina display).

<img width="120" alt="Screen Shot 2019-11-20 at 16 23 17" src="https://user-images.githubusercontent.com/3276087/69283509-2d480400-0bb2-11ea-96be-b74a4e69f1df.png">

Accessibility
---

I was able to interact with the welcome guide using only my keyboard 👍

I also did a quick test with Voice Over and had a couple of issues:

1. 
The page control buttons could use an `aria-label` so it's communicated to assistive technology what they're for. I was expecting to hear something like "page 1 of 3".

I'm thinking these could be grouped on a `ul` so they're part of the same group. Following suggestions on [how to markup content sliders](https://inclusive-components.design/a-content-slider/) from [the Inclusive Components pattern library](https://inclusive-components.design), it could be something like:

```
<ul aria-label="Welcome guide controls">
    <li><button aria-label="Page 1 of 3"></button></li>
    <li><button aria-label="Page 2 of 3"></button></li>
    <li><button aria-label="Page 3 of 3"></button></li>
</ul>
```
2. 
I don't think the images need an `alt` description, because the accompanying text does a good job of describing each page/topic. That being said, they should have an empty `alt = ""` [so that they're ignored by AT](https://www.w3.org/WAI/tutorials/images/decorative/).

Overall, this is looking great! Let me know what you think of my comments and suggestions.

enriquesanchez commented:
@noisysocks where are you getting the images from? Let me know if you need help with Mark's suggestions on alignment and size.

noisysocks commented:
Thanks @mapk and @enriquesanchez! I'll work through your suggestions.

> @noisysocks where are you getting the images from? Let me know if you need help with Mark's suggestions on alignment and size.

I would absolutely love it if you could update the images for me! I got the SVGs from @jasmussen, who I think exported them from the 'Tips Iteration' project in the WordPress.org Figma team.

enriquesanchez commented:
Hey @noisysocks! (I never get tired of your username 😃)

Here are the new SVGs with Mark's suggested changes:

[illustrations.zip](https://github.com/WordPress/gutenberg/files/3876246/illustrations.zip)

enriquesanchez commented:
Wait, I noticed the add block icon on 'illustration 3.svg` was still showing some weird artifacts. 

Here's a cleaner version:
[illustration 3_.svg.zip](https://github.com/WordPress/gutenberg/files/3876266/illustration.3_.svg.zip)

noisysocks committed:
Welcome guide: Update images

noisysocks committed:
Welcome guide: Capitalise 'Block Editor' and 'Block Library'

noisysocks committed:
Welcome guide: Position close button 24px from edge

noisysocks committed:
Welcome guide: Improve a11y of page controls

noisysocks commented:
> 1. First image needs to be centered better.
> 2. First two graphics should be equal width.
> 5. This icon can probably use some visual cleanup.

👍 I updated the images in d8cee821dc14c4c4dbb9494534f2b0411e97422e. Thanks for fixing those up @enriquesanchez!

> 3. Should we capitalize certain things?

👍 I capitalised 'Block Editor' and 'Block Library' in 745fdd44ceb3821af508a90ebbd35411a9107f8c.

> 4. Second screen text addition

I'll defer to you all on this one. Let me know what you'd like the copy to be and I can update it.

> Would it be possible to align the close button so it has the same space around like the previous and next/get-started buttons?

I updated the alignment in 47c2879888c27ee743c1b7189197977c210b20fc, though I wonder if doing this makes the X button look too much like 'content' instead of 'chrome'? Let me know what you think.

<img width="744" alt="Screen Shot 2019-11-22 at 11 42 11" src="https://user-images.githubusercontent.com/612155/69421957-b5e5b200-0d22-11ea-8e22-d2659666acef.png">

> The page control buttons could use an aria-label so it's communicated to assistive technology what they're for. I was expecting to hear something like "page 1 of 3".

👍 I updated the page controls to use the markup you suggested in 92e1773ddd73af1851c480ece84af2f8c709d565.

> I don't think the images need an alt description, because the accompanying text does a good job of describing each page/topic. That being said, they should have an empty alt = "" so that they're ignored by AT.

The images already have an empty `alt=""` attribute.

noisysocks committed:
Welcome guide: Add unit tests

noisysocks committed:
Welcome guide: Add E2E tests

noisysocks committed:
Welcome guide: Mark Guide as __experimental, as we may want to move it to @wordpress/components

noisysocks committed:
Welcome guide: Add README.md

noisysocks commented:
> > Separately, it looks like #12632 would still be effected by this tips redesign. AFAICT, the close button here is dismissing all guides instead of just the current "guide".
> Thanks for pointing me to #12632. I'll look into moving away from areTipsEnabled(). My preference is to make Guide a dumb component and leave it up to the plugin developer to decide when to show it.

Let's look at making the welcome guide use something other than `areTipsEnabled` in a separate PR. There are some details to consider there and I want us to have the opportunity to properly think it through. We also need to consider whether or not we want to maintain or deprecate the `@wordpress/nux` package. #12632 can track that work.

> YES! Adding this to @wordpress/components sounds reasonable to me. Allowing other developers to use it could be amazing.

I've left `Guide` in the `@wordpress/nux` package for now because it requires `isViewportMatch` to determine whether or not to display the 'Get started' button inline with the content. Let's consider moving `Guide` to `@wordpress/components` in a follow-up PR—I've marked `Guide` as experimental for now so that we can do this.

noisysocks committed:
Welcome guide: Rename WelcomeGuideModal to WelcomeGuide

noisysocks committed:
Welcome guide: Update fixtures

noisysocks commented:
This is ready for review! ❤️

(Seeing an unrelated test failure—have restarted the job.)

mapk reviewed (approved):
Just tested. It looks good to me! :shipit: 

![nux 2019-11-27 14_57_48](https://user-images.githubusercontent.com/617986/69764789-6dfdcb00-1126-11ea-9526-38c4a24fb384.gif)

enriquesanchez commented:
Hey @noisysocks! My dev environment is borked right now, but I well review this as soon as I get it up and running.

enriquesanchez commented:
Just tested and it looks good! Thanks for working on the page controls, it's now clear what they are for when using Voice Over 👍 🎉

noisysocks commented:
Thanks @mapk and @enriquesanchez! Could someone please give this a code review?

youknowriad commented on packages/edit-post/src/components/welcome-guide/images.js:
```diff
@@ -0,0 +1,36 @@
+		src="data:image/svg+xml;charset=utf8,%3Csvg width='18' height='18' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M8.824 0C3.97 0 0 3.97 0 8.824c0 4.853 3.97 8.824 8.824 8.824 4.853 0 8.824-3.971 8.824-8.824S13.677 0 8.824 0zM7.94 5.294v2.647H5.294v1.765h2.647v2.647h1.765V9.706h2.647V7.941H9.706V5.294H7.941zm-6.176 3.53c0 3.882 3.176 7.059 7. ...
+		{ ...props }
+	/>
+);
```
Did you check the impact on the bundle size of the script?

noisysocks replied:
`master`:

```
./build/nux/index.js    7.46 KiB      34  [emitted]         nux
```

`add/welcome-guide`:

```
 ./build/nux/index.js    10.9 KiB      34  [emitted]         nux
```

So an extra 3.44 KB. What do you think? Happy to move them somewhere else.

I was thinking ideally I'd like to have a SVG loader set up in Webpack so that we can `import()` it in a lazy fashion. Maybe I can add a note to https://github.com/WordPress/gutenberg/issues/2768.

youknowriad replied:
Taking about `edit-post` package.

noisysocks replied:
🤦‍♂️ ich bin dumm.

`master`:

```
./build/edit-post/index.js    83.9 KiB      20  [emitted]         editPost
```

`add/welcome-guide`:

```
./build/edit-post/index.js    98.2 KiB      20  [emitted]         editPost
```

14.3 KB.

youknowriad replied:
That's less than I expected, so I think it's acceptable.

youknowriad replied:
je bent een ster

youknowriad commented on packages/edit-post/src/components/welcome-guide/index.js:
```diff
@@ -0,0 +1,71 @@
+			onFinish={ disableTips }
+		>
+
+			<Guide.Page className="edit-post-welcome-guide__page">
```
`Guide.Page` this is not a pattern we used that much, we tend to favor `GuidePage` instead. (PanelBody, ...) 
I'm not strongly against it but we should be consistent I think.

noisysocks replied:
👌 ec96e0e7af8c2b82be8174bd0f4d2fe5b9c8afbc

youknowriad commented on packages/edit-post/src/plugins/index.js:
```diff
@@ -29,6 +30,7 @@ registerPlugin( 'edit-post', {
 								{ __( 'Manage All Reusable Blocks' ) }
 							</MenuItem>
 							<KeyboardShortcutsHelpMenuItem onSelect={ onClose } />
+							<WelcomeGuideMenuItem onSelect={ onClose } />
```
not specific to this PR but we're using the "More Menu" as a place where we just add things we're not sure where to add. Some design thinking here could be good. cc @jasmussen @mapk

gziolo replied:
`onSelect` can be omitted, it's not used and probably shouldn't be so keyboard users could experience the focus moved back to this item when they close the welcome guide.

noisysocks replied:
Agree that design input would be good.

I think ideally a button like this would go into a _Help_ menu. For example, Figma has this nice floating _Help_ button in the bottom right of the screen:

<img width="180" alt="Screen Shot 2019-11-28 at 11 58 02" src="https://user-images.githubusercontent.com/612155/69800976-9b09b680-11d6-11ea-805f-8d989280fd7e.png">

Until we have something like that, though, I think it makes sense to put this in the _More_ menu.

---

@gziolo: Good catch! Fixed in 588c2d53b.

youknowriad commented on packages/nux/src/components/guide/README.md:
```diff
@@ -0,0 +1,56 @@
+Guide
```
I wonder if this could be just a `@wordpress/components` component instead

noisysocks replied:
👌 moved `Guide` over to `@wordpress/components` in 78104f3db0190d6fac295933d44bee00a5d15c9f.

youknowriad commented on packages/nux/src/components/guide/index.js:
```diff
@@ -0,0 +1,116 @@
+}
+
+const EnhancedGuide = withSelect( ( select ) => ( {
+	isMobile: select( 'core/viewport' ).isViewportMatch( '< small' ),
```
I think we have specific HoC to do exactly that for you instead of using the store directly. `withViewpostMatch`

noisysocks replied:
Ah, perfect! Thoughts on adding `@wordpress/viewport` as a dependency to `@wordpress/components` so that we can move `Guide` into `@wordpress/components`?

youknowriad replied:
I do remember that we discussed the same thing a long time ago. I think I'm fine with it (what I don't like is the indirect dependency on the data module). That said, I feel like the whole viewport package could be just a `useViewportMatch` hook in the compose package.

noisysocks replied:
I changed `Guide` to use `useMediaQuery` and moved the component to `@wordpress/components` in 78104f3db0190d6fac295933d44bee00a5d15c9f. I like the idea of deprecating `@wordpress/viewport` in lieu of a hook in `@wordpress/compose`, but let's look into that separately.

youknowriad commented on packages/nux/src/components/guide/test/__snapshots__/index.js.snap:
```diff
@@ -0,0 +1,41 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Guide renders when there are pages 1`] = `
```
I don't like snapshot testing personally, I feel we just end up updating snapshots when the component change.

youknowriad replied:
We could instead look for the Pages component and just check that their length is one.

noisysocks replied:
I like that having to update the snapshot when a component changes means that the reviewer can see exactly how the markup has changed in the PR's diff.

But you're the boss! 😀 removed in 809d0861f.

youknowriad reviewed (commented):
Awesome work here.

gziolo commented on packages/edit-post/src/components/welcome-guide/index.js:
```diff
@@ -0,0 +1,71 @@
+	return (
+		<Guide
+			className="edit-post-welcome-guide"
+			finishButtonText={ __( 'Get started' ) }
```
I find it interesting that when you click the finish button you actually get started 😄 

Did you consider using some alternatives?

noisysocks replied:
I think `finishButton` is correct. It's the button that, when clicked, marks the guide as _finished_ and calls `onFinish`. 

In our case, `WelcomeGuide` shows the `Guide` to brand-new users and so _finishing the guide_ is equivalent to _getting started with the editor_. But `Guide` is generic and could be used in other cases e.g. a plugin tutorial that displays a guide when a _Help_ button is clicked.

gziolo commented on packages/nux/src/components/guide/page-control.js:
```diff
@@ -0,0 +1,33 @@
+
+export default function PageControl( { currentPage, numberOfPages, setCurrentPage } ) {
+	return (
+		<ul className="nux-guide__page-control" aria-label={ __( 'Guide controls' ) }>
```
See:
https://github.com/WordPress/gutenberg/blob/853b3cb100469f9d71bad93bb38ece5210dae0b2/packages/edit-post/src/components/keyboard-shortcut-help-modal/index.js#L44-L49

Do we need it here as well? We should finally abstract it away, as it pops up way too often.

noisysocks replied:
Will defer to @enriquesanchez here. He said in https://github.com/WordPress/gutenberg/pull/18041#issuecomment-559300177 that this worked in VoiceOver, so I'm not sure that we need to change the markup.

gziolo replied:
It would be great if they fix it.

enriquesanchez replied:
Are you referring to the `ul`?

gziolo replied:
Yes, it was problematic in other places so we had to explicitly provide the `role="list"`.

gziolo commented on packages/nux/src/components/guide/test/finish-button.js:
```diff
@@ -0,0 +1,49 @@
+	it( 'calls onClick when the button is clicked', () => {
+		const onClick = jest.fn();
+		const wrapper = shallow( <FinishButton onClick={ onClick } /> );
+		wrapper.find( Button ).prop( 'onClick' )();
```
I think it tests the internals of React in the current form. It doesn't ensure that `onClick` is properly wired with the DOM.

noisysocks replied:
It's a unit test so we should only test the part that concerns `FinishButton`. In this case, we're testing that the `Button` component's `onClick` prop is wired to the `FinishButton` component's `onClick` prop. It's up to the `Button` component and its unit tests to ensure that the actual DOM event is handled.

(This test caught a bug while I was refactoring things in 78104f3db0190d6fac295933d44bee00a5d15c9f, too! 😀)

noisysocks committed:
Welcome guide: Rename Guide.Page to GuidePage

noisysocks committed:
Welcome guide: Remove unnecessary onSelect prop

gziolo commented on packages/edit-post/src/components/welcome-guide/images.js:
```diff
@@ -0,0 +1,36 @@
+
+export const CanvasImage = ( props ) => (
+	<img
+		alt=""
```
Shouldn't those `alt` attributes be filled in for those who use screen readers?

noisysocks replied:
From https://github.com/WordPress/gutenberg/pull/18041#issuecomment-556508056:

> I don't think the images need an alt description, because the accompanying text does a good job of describing each page/topic. That being said, they should have an empty alt = "" so that they're ignored by AT.

gziolo commented:
> Just tested. It looks good to me! :shipit:
> 
> ![nux 2019-11-27 14_57_48](https://user-images.githubusercontent.com/617986/69764789-6dfdcb00-1126-11ea-9526-38c4a24fb384.gif)

Great work! 🥇

noisysocks committed:
Welcome guide: Move Guide to @wordpress/components

noisysocks committed:
Welcome guide: Remove snapshot test

youknowriad commented on packages/components/src/guide/index.js:
```diff
@@ -0,0 +1,108 @@
+import FinishButton from './finish-button';
+
+export default function Guide( { children, className, finishButtonText, onFinish } ) {
+	const isMobile = useMediaQuery( '(max-width: 600px)' );
```
Would be good to abstract the threshods in a `useViewport` hook like we do in the viewports package but this is probably a separate PR I think.

youknowriad replied:
The was `isMobile` is used in this component though makes me wonder if it should be just a CSS media query :)

noisysocks replied:
I had thought that using CSS here didn't play nicely with the auto-focus behaviour that `FinishButton` has, but it seems okay now! Switched to CSS in 9b5e1dd17.

(We should still deprecate `@wordpress/viewport` though—it's a good idea.)

youknowriad commented on packages/components/src/index.js:
```diff
@@ -33,6 +33,8 @@ export { default as FormFileUpload } from './form-file-upload';
 export { default as FormTokenField } from './form-token-field';
 export { default as __experimentalGradientPicker } from './gradient-picker';
+export { default as Guide } from './guide';
+export { default as GuidePage } from './guide/page';
```
Storybook stories? :)

noisysocks replied:
how do I do that 🙈

youknowriad replied:
haha :) look at others `stories` folder (I think adding the folder and exporting a function from the index should be enough)

noisysocks replied:
Done! f8c494d80

<img width="1552" alt="Screen Shot 2019-11-28 at 14 42 13" src="https://user-images.githubusercontent.com/612155/69811002-6bb27400-11ed-11ea-999f-9089555924df.png">

youknowriad reviewed (approved):
This is looking good to me. I'd appreciate a quick sanity check from others (cc @ItsJonQ) whether the new components make sense in our design system :)

gziolo commented:
> I'd appreciate a quick sanity check from others (cc @ItsJonQ) whether the new components make sense in our design system :)

Does it make sense to open a follow-up and refactor `Guide` and/or `GuidePage` to be composed of `Card*` components?

noisysocks committed:
Welcome guide: Use CSS to hide and show finish buton

youknowriad commented on packages/components/src/guide/style.scss:
```diff
@@ -104,5 +104,15 @@
 	&.components-guide__finish-button {
 		right: 0;
+
+		@media (max-width: $break-small) {
```
Can we instead invert the style (mobile first) and use our mixins?

noisysocks replied:
I find it odd setting `display: none` only to then set `display: unset`. But happy to use the mixin if you prefer :)

youknowriad replied:
I don't like it a lot either. but you know we have to follow the gurus of mobile-first approach and stay consistent in our codebase :)

noisysocks replied:
I want you to be happy, Riad ❤️ 53974ef3c

noisysocks committed:
Welcome guide: Add a storybook

noisysocks committed:
Welcome guide: Use break-small() mixin

noisysocks commented:
Thanks for reviewing @youknowriad @gziolo!

Just waiting on some final feedback from @ItsJonQ (about https://github.com/WordPress/gutenberg/pull/18041#pullrequestreview-324270220) and @enriquesanchez (about https://github.com/WordPress/gutenberg/pull/18041#discussion_r351761572) and we should be good to merge.

noisysocks committed:
Welcome guide: Generate storybook snapshot

noisysocks committed:
Welcome guide: Update @wordpress/components changelog

noisysocks committed:
E2E tests: Reload page after disabling welcome guide / tips

noisysocks commented on packages/e2e-test-utils/src/create-new-post.js:
```diff
@@ -28,12 +28,13 @@ export async function createNewPost( {
+			const action = _enableTips ? 'enableTips' : 'disableTips';
+			wp.data.dispatch( 'core/nux' )[ action ]();
+		}, enableTips );
 		await page.reload();
```
Rebasing https://github.com/WordPress/gutenberg/pull/18775 into this PR caused tests that rely on the post title receiving initial focus to fail. Reloading the page after we first disable the welcome guide seems to fix.

cc. @youknowriad

noisysocks committed:
Welcome guide: Close guide when user clicks on overlay

joedolson commented:
The modal is not labelled. This is in part a problem with this modal, but should probably be corrected in the component: the Modal component allows the component to be misused.

We should enforce a label for the modal element with role=dialog within the e2e tests, because all modals must be labelled. Additionally, the modal component should enforce 'role="document"' on the content of all modal dialogs. This isn't strictly required on all modals, but the component can't know what type of content a modal will contain, this will give the best overall experience.

There are usability problems with having this modal open automatically on page load.

1) As it is (unlabeled), it's going to be confusing for some users.
2) It disrupts the flow for users who wish to use Gutenberg immediately.
3) Users who reflexively dismiss modals will not easily be able to get it back.

One possible way that this could be made available is to add a prominent button for users to launch the Welcome Modal if they are detected as new Gutenberg users. The modal, in that case, should have two cancel actions: one that simply closes the modal, and one that dismisses it persistently.

As a second issue, the method for assessing whether a user is "new" to Gutenberg is significantly flawed: the usage of local storage for this means that a user will be new on every device they use, on every browser they use, and on every new site they launch. Secondarily, because the detection appears to be based on the positive presence of 'modal dismissed' setting in local storage, it will essentially be impossible to permanently disable it; anytime the local storage is flushed it will come back. (See https://github.com/WordPress/gutenberg/issues/15105)

aduth commented on packages/components/src/guide/style.scss:
```diff
@@ -0,0 +1,119 @@
+		}
+	}
+
+	&__container {
```
While I think this convention certainly makes things more concise, in practice I find it adds difficulty to a very common use-case of searching for a class name in the codebase. To try to search for "components-guide__container" would not yield this as a result. While it requires a couple more characters to type, I personally prefer the fully expanded form in source:

```scss
.components-guide__container {
```

aduth commented on packages/components/src/guide/finish-button.js:
```diff
@@ -0,0 +1,33 @@
+	// Focus the button on mount if nothing else is focused. This prevents a
+	// focus loss when the 'Next' button is swapped out.
+	useLayoutEffect( () => {
+		if ( document.activeElement === document.body ) {
```
In #20594, I am encountering an issue where we make certain assumptions about what it means for "no focus" to exist. It appears that the value of `activeElement` may differ depending on browser:

>When there is no selection, the active element is the page's `<body>` or `null`.

https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/activeElement

I mention it here, since this may not cover all focus losses, notably those in Internet Explorer.

![image](https://user-images.githubusercontent.com/1779930/75707343-b92afb00-5c8c-11ea-8084-a4b9571e0531.png)

It might be something where we want to provide some utility, e.g. `wp.dom.hasActiveElement`, since it seems to be an easy issue to overlook.

aduth replied:
Fix at #20599
